selinux: Fix the controller to ignore finished pods

jsafrane opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Updated the SELinux warning controller to ignore completed Pods (both Succeeded and Failed). Kubelet is umounting volumes from those Pods by default, so they cannot conflict with any other Pod. To the controller it's the same as if they were deleted.

#### Which issue(s) this PR is related to:
Fixes #135628

#### Special notes for your reviewer:
See individual commits. There is some minor refactoring of related unit tests + new e2e tests.

The new tests revealed another bug - a Pod with no SELinux label must be treated as with `SELinuxChangePolicy: Recursive`. Kubelet already does that - it does not have any label to file into `mount -o context=<label>`. The controller is fixed in the last commit in this PR.

#### Does this PR introduce a user-facing change?
```release-note
Fixed SELinux warning controller not to emit events for completed pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
selinux: Do not report conflits with finished pods

When a Pod reaches its final state (Succeeded or Failed), its volumes are
getting unmounted and therefore their SELinux mount option will not
conflict with any other pod.

Let the SELinux controller monitor "pod updated" events to see the pod is
finished

jsafrane committed:
selinux: add e2e test with a completed pod

Add a test that checks the SELinux controller does not report conflicts
with Succeeded pods.

jsafrane committed:
refactoring: use a common function to enqueue Pod

addPod and deletePod have the same implementation, merge them into
enqueuePod

jsafrane committed:
Rework unit tests to builder pattern

jsafrane committed:
Add new unit tests

jsafrane committed:
Fix policy of Pods with unknown SELinux label

Reset SELinuxChangePolicy of Pods that have no SELinux label set to
Recursive. Kubelet cannot mount with `-o context=<label>`, if the label is
not known.

This fixes the e2e test error revealed by the previous commit - it changed the
e2e test to check for events when no events are expected and it found a
warning about a Pod with no label, but MountOption policy.

gnufied commented on pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:
```diff
@@ -187,6 +187,28 @@ func (c *Controller) enqueuePod(_ klog.Logger, obj interface{}) {
+		return
+	}
+	if newPod.Status.Phase != v1.PodFailed && newPod.Status.Phase != v1.PodSucceeded {
+		return
```
This is basically an optimization right? The completed pods will anyways be skipped during processing below?

jsafrane replied:
I added a comment + squashed it into the commit that introduces this check.

jsafrane committed:
Use only enqueuePod to add pods to the controller queue

enqueuePod already creates the right key for a pod, it's better to reuse it
than copy the code around.

jsafrane committed:
Added e2e tests with disabled SELinux

Added few tests with a CSI driver that does not support SELinux and has it
disabled in its CSIDriver instance

jsafrane committed:
Add unit test with CSIDriver.SELinuxMount=false

Add unit test with a volume plugin that does not support SELinux. That
simulates a CSi driver whose spec.SELinuxMount is empty or false.

This requires a little refactoring, each unit test now has a flag if it
runs with a volume plugin that supports SELinux.

jsafrane commented on pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:
```diff
@@ -489,8 +507,15 @@ func (c *Controller) syncVolume(logger klog.Logger, pod *v1.Pod, spec *volume.Sp
+	if !pluginSupportsSELinuxContextMount && changePolicy != v1.SELinuxChangePolicyRecursive {
+		logger.V(5).Info("Volume does not support SELinux context mount, setting changePolicy to Recursive", "pod", klog.KObj(pod), "volume", spec.Name())
+		changePolicy = v1.SELinuxChangePolicyRecursive
+	}
```
added a single unit test + some unit test refactoring as a new commit

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux
